odoo/websocket.py

```
@@ -0,0 +1,959 @@
+            return False
+        return other.opcode == self.opcode and self.payload == other.payload
+
+    def __lt__(self, other):
+        if not isinstance(other, Frame):
+            raise TypeError("< only supported between instances of 'Frame'")
+        if (
+            self.opcode in HEARTBEAT_OP and
+            other.opcode in HEARTBEAT_OP or
+            self.opcode in DATA_OP and other.opcode in DATA_OP
+        ):
+            return self._send_order < other._send_order
+        return self.opcode in HEARTBEAT_OP
+
+
+class CloseFrame(Frame):
+    def __init__(self, code, reason):
+        if code not in VALID_CLOSE_CODES and code not in RESERVED_CLOSE_CODES:
+            raise InvalidCloseCodeException(code)
+        payload = struct.pack('!H', code)
+        if reason:
+            payload += reason.encode('utf-8')
+        self.code = code
+        self.reason = reason
+        super().__init__(Opcode.CLOSE, payload)
+
+
+class Websocket:
+    _instances = WeakSet()
+    _event_callbacks = defaultdict(set)
+    # Maximum size for a message, whether it is sent as one frame or
+    # many fragmented ones.
+    MESSAGE_MAX_SIZE = 2 ** 40
+    # Proxies usually close a connection after 1 minute of inactivity.
+    # Therefore, a PING frame have to be sent if no frame is either sent
+    # or received within CONNECTION_TIMEOUT - 15 seconds.
+    CONNECTION_TIMEOUT = 60
+    INACTIVITY_TIMEOUT = CONNECTION_TIMEOUT - 15
+    # How many requests can be made in excess of the given rate.
+    RL_BURST = config.get('websocket_rate_limit_burst', 10)
+    # How many seconds between each request.
+    RL_DELAY = config.get('websocket_rate_limit_delay', 0.2)
+
+    def __init__(self, socket):
+        self._socket = socket
+        self._close_sent = False
+        self._outgoing_frame_queue = PollablePriorityQueue()
+        self._timeout_manager = TimeoutManager()
+        # Used for rate limiting.
+        self._inncoming_frame_timestamps = deque()
```

```suggestion
        self._incoming_frame_timestamps = deque(maxlen=type(self).RL_BURST)
```

The rate-limit test suite should fail for the missing maxlen. In case it passes, you should add more tests :p

*It all started at Mander-Alcord, in the traveler's inn. Our hero was looking for adventurers to help him in his quest.*